qiskit_machine_learning/algorithms/objective_functions.py

```diff
@@ -0,0 +1,158 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""An abstract objective function definition and common objective functions suitable
+for classifiers/regressors."""
+
+from abc import abstractmethod
+from typing import Optional, Union
+
+import numpy as np
+from sparse import SparseArray
+
+from qiskit_machine_learning.neural_networks import NeuralNetwork
+from qiskit_machine_learning.utils.loss_functions import Loss
+
+
+class ObjectiveFunction:
+    """An abstract objective function. Provides methods for computing objective value and
+    gradients for forward and backward passes."""
+
+    # pylint: disable=invalid-name
+    def __init__(
+        self, X: np.ndarray, y: np.ndarray, neural_network: NeuralNetwork, loss: Loss
+    ) -> None:
+        """
+        Args:
+            X: The input data.
+            y: The target values.
+            neural_network: An instance of an quantum neural network to be used by this
+                objective function.
+            loss: A target loss function to be used in training.
+        """
+        super().__init__()
+        self._X = X
+        self._y = y
+        self._neural_network = neural_network
+        self._loss = loss
+        self._last_forward_id: Optional[str] = None
+        self._last_forward: Optional[Union[np.ndarray, SparseArray]] = None
+
+    @abstractmethod
+    def objective(self, weights: np.ndarray) -> float:
+        """Computes the value of this objective function given weights.
+
+        Args:
+            weights: an array of weights to be used in the objective function.
+
+        Returns:
+            Value of the function.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def gradient(self, weights: np.ndarray) -> np.ndarray:
+        """Computes gradients of this objective function given weights.
+
+        Args:
+            weights: an array of weights to be used in the objective function.
+
+        Returns:
+            Gradients of the function.
+        """
+        raise NotImplementedError
+
+    def _neural_network_forward(self, weights: np.ndarray) -> Union[np.ndarray, SparseArray]:
+        """
+        Computes and caches the results of the forward pass. Cached values may be re-used in
+        gradient computation.
+
+        Args:
+            weights: an array of weights to be used in the forward pass.
+
+        Returns:
+            The result of the neural network.
+        """
+        data_id = str(id(weights)) + str(id(self._X))
```

Comment:
I'd expect a comparison by id to be faster than a direct comparison of arrays. Well, it is not the case now, we have just a few qubits, but in case of classic ML, arrays maybe huge.
It is also a good point whether we need to add `_X` to the comparison. Honestly, I added it only for safety reasons. In general, a comparison of weights is good enough, but just in case. So, if you don't mind I'd keep it.

Reply:
Sure, I was just trying to understand things from whats there. Whether its worth a comment in the code to that effect, i.e. safety just in case, to help any other new readers of the code I'll leave to you.